### PR TITLE
converted TeachAPI mixin to higher order component 

### PR DESCRIPTION
**Pull request by Pomax**
fixes https://github.com/mozilla/teach.mozilla.org/issues/1556

Replaced the teachapi mixin with a Higher Order Component, and updated all dependents and tests accordingly.

For some reason if I run `npm test`, the sidebar times out for me, but if I run `npm start` and then check http://localhost:8008/test, the sidebar passes pretty much immediately, so that's something odd...

Note that this is based on the https://github.com/mozilla/teach.mozilla.org/pull/1607, so that needs to land first, so this PR can be rebased before landing.

### CODE REVIEW

**Comment by mmmavis** [comment]
@Pomax could you update this PR so it doesn't have merge conflicts? :grin:

**Comment by Pomax** [comment]
@mmmavis: rebased

**Comment by mmmavis** [comment]
@Pomax 

> For some reason if I run npm test, the sidebar times out for me, but if I run npm start and then check http://localhost:8008/test, the sidebar passes pretty much immediately, so that's something odd...

I think it's probably cuz staging teach-api lives on Heroku and needs time to wake up from inactivity?

File: components/login/LogoutLink.jsx [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+    return {
+      origin: config.ORIGIN
+    };
+  },
+  render: function() {
+    var callbackURL = this.props.origin + this.context.router.getCurrentPathname();
+    var loginBaseURL = this.props.loginBaseURL;
+    var href = loginBaseURL + '/auth/oauth2/logout?callback=' + encodeURIComponent(callbackURL);
+    var props = _.extend({}, this.props, { href: href });
+    return <a {...props}>{this.props.children}</a>;
```
Comment (mmmavis) [inline components/login/LogoutLink.jsx]: Not sure if we want to track log out actions too?

Comment (Pomax) [inline components/login/LogoutLink.jsx]: track in what way, ga.event?

File: components/modal-clubs.jsx [hunk 2]
```diff
@@ -165,127 +168,147 @@ var ModalAddOrChangeYourClub = React.createClass({
...
+         : <h2>Your club has been changed.</h2>}
+        <button className="btn btn-block"
+         onClick={this.handleSuccessClick}>
+          Take Me To My Club
+        </button>
+      </div>
+    );
+  },
+
+  render: function() {
```
Comment (mmmavis) [inline components/modal-clubs.jsx]: love this clean render function

File: hoc/with-teach-api.jsx [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+//   var Foo = React.createClass({
+//     ...
+//     render: function() {
+//       return <div>I am {this.props.teachAPI.getUserName()}</div>;
+//     }
+//   });
+//
+//   module.exports = withTeachAPI(Foo);
+//
+// The mixin will ensure that the handleLogout() method is called whenever
```
Comment (mmmavis) [inline hoc/with-teach-api.jsx]: Could we tweak the wording since it's not a "mixin" anymore :hamster:

Comment (Pomax) [inline hoc/with-teach-api.jsx]: yeah, good call

**Comment by mmmavis** [comment]
R+++++ Really nice and clean rewrite @Pomax ! Thanks! :+1: :+1: :+1:

File: components/login/LogoutLink.jsx (hunk 1 above)
Comment (mmmavis) [inline components/login/LogoutLink.jsx]: Oh yea, like using the `<OutboundLink>` component.

**Pull request merged**